src/MongoDB.EntityFrameworkCore/Storage/MongoDatabaseWrapper.cs

```diff
@@ -217,4 +148,148 @@ private static string GetId(IUpdateEntry entry)
             _ => throw new NotSupportedException($"Unknown type for _id property on entity type ${entry.EntityType.DisplayName()}")
         };
     }
+
+    private IEnumerable<MongoUpdate> ConvertUpdateEntriesToMongoUpdates(IEnumerable<IUpdateEntry> entries)
+    {
+        return
+            GetAllChangedRootEntries(entries)
+            .Select(ConvertUpdateEntryToMongoUpdate)
+            .Where(x => x != null)
+            .Cast<MongoUpdate>();
+    }
+
+    private MongoUpdate? ConvertUpdateEntryToMongoUpdate(IUpdateEntry entry)
+    {
+        var collectionName = entry.EntityType.GetCollectionName();
+        var id = GetId(entry);
+
+        return entry.EntityState switch
+        {
+            EntityState.Added => ConvertAddedEntryToMongoUpdate(collectionName, id, entry),
+            EntityState.Deleted => ConvertDeletedEntryToMongoUpdate(collectionName, id, entry),
+            EntityState.Detached => null,
+            EntityState.Modified => ConvertModifiedEntryToMongoUpdate(collectionName, id, entry),
+            EntityState.Unchanged => null,
+            _ => throw new NotSupportedException($"Unexpected entity state: {entry.EntityState}.")
+        };
+    }
+
+    private MongoUpdate ConvertAddedEntryToMongoUpdate(string collectionName, string id, IUpdateEntry entry)
+    {
+        throw new NotImplementedException(); // EF-17
+    }
+
+    private MongoUpdate ConvertDeletedEntryToMongoUpdate(string collectionName, string id, IUpdateEntry entry)
+    {
+        var filter = new BsonDocument("_id", id);
+        var model = new DeleteOneModel<BsonDocument>(filter);
+        return new MongoUpdate(collectionName, model);
+    }
+
+    private MongoUpdate ConvertModifiedEntryToMongoUpdate(string collectionName, string id, IUpdateEntry entry)
+    {
+        throw new NotImplementedException(); // EF-15
+    }
+
+    private long SaveMongoUpdates(IEnumerable<MongoUpdate> updates)
+    {
+        var database = _mongoClient.Database;
+        var client = database.Client;
+        using var session = client.StartSession();
+        return session.WithTransaction((session, cancellationToken) => SaveMongoUpdates(session, database, updates));
```

Comment:
We don't have to use a transaction if we decide not to.

But this shows that it is trivial to use a transaction for the whole batch update if we want to.

Reply:
Let's keep the transactions.

Reply:
will this also work in cases where it's just a single node? (not a replica set)